index.bs

```diff
@@ -241,6 +243,66 @@ The <dfn attribute for=FileSystemHandle>name</dfn> getter steps are to return [=
   :: Returns true if |handle1| and |handle2| represent the same file or directory.
 </div>
 
+### The {{FileSystemHandle/remove()}} method ### {#api-filesystemhandle-remove}
+
+<div class="note domintro">
+  : await |handle| . {{FileSystemHandle/remove()|remove}}()
+  : await |handle| . {{FileSystemHandle/remove()|remove}}({ {{FileSystemRemoveOptions/recursive}}: false })
+  :: Attempts to remove the entry represented by |handle| from the underlying file system.
+
+     Attempting to delete a file or directory that does not exist is considered success,
+     while attempting to delete a non-empty directory will result in a promise rejection.
+
+  : await |handle| . {{FileSystemHandle/remove()|remove}}({ {{FileSystemRemoveOptions/recursive}}: true })
+  :: Attempts to remove the entry represented by |handle| from the underlying file system.
+
+     If the entry is a directory, its contents will also be deleted recursively.
+
+     Attempting to delete a file or directory that does not exist is considered success.
+
+     For files or directories with multiple hardlinks or symlinks, the entry which is deleted
+     is the entry corresponding to the path which was used to obtain the handle.
+</div>
+
+<div algorithm>
+The <dfn method for=FileSystemHandle>remove(|options|)</dfn> method steps are:
+
+1. Let |result| be [=a new promise=].
+1. Run these steps [=in parallel=]:
+  1. Let |entry| be <b>[=this=]</b>'s [=FileSystemHandle/entry=].
+  1. Let |access| be the result of running |entry|'s
+     [=entry/request access=] given "`readwrite`".
+  1. If |access| is not "{{PermissionState/granted}}", reject |result| with a
+     "{{NotAllowedError}}" {{DOMException}} and abort.
+
+  1. Let |lockResult| be the result of [=file entry/lock/take|taking a lock=]
+     with "`exclusive`" on |entry|.
+  1. If |lockResult| is false, [=reject=] |result| with a
+     "{{NoModificationAllowedError}}" {{DOMException}} and abort.
+
+  1. If |entry| is a [=directory entry=]:
+    1. If |entry| is not an empty directory and |options|.{{FileSystemRemoveOptions/recursive}} is `false`:
+      1. [=/Reject=] |result| with an "{{InvalidModificationError}}" {{DOMException}} and abort.
+   1. If |entry| is the [=root directory of the origin private file system=]:
+      1. [=/Reject=] |result| with a "{{NoModificationAllowedError}}" {{DOMException}} and abort.
```

Comment:
Ooh, so this is covered. It seems kinda nice to allow this, no?

Reply:
It seems useful to allow for recursive removal of all children of the root - but can we allow removal of the root itself? Could we specify that the root re-creates itself on the next call of `navigator.storage.getDirectory()`?

(outside the scope of this discussion, but if you really want to wipe the entire OPFS that seems like a use case for the Storage Buckets API)

Reply:
It seems fine to allow removing root directory as it provides an easy way to clear OPFS. `navigator.storage.getDirectory()` can create the root directory again (which I think is the current behavior, i.e. we have to create the directory when clients calls `navigator.storage.getDirectory()` for the first time, or after users delete website data.)

Reply:
@a-sully I suppose we could leave the root directory itself in specification-wise if that simplifies things as there's no state associated with it that would make that observable (you won't get the same object back representing it, for instance).

But yeah, as @szewai says I would expect it to act as if it was removed and the next call creates a new one.

Reply:
I was worried about how this interacts with https://github.com/whatwg/fs/issues/39, but I guess with directories, unlike with files, there's no way to re-create the directory once it's been removed without access to the parent (which is impossible in this case). So yeah, specifying that the next invocation of `navigator.storage.getDirectory()` creates a new OPFS seems reasonable. 

This also solves an existing problem in the Chromium implementation (mentioned in https://github.com/whatwg/fs/issues/59) where handles in the OPFS could become suddenly valid if the path is re-created from underneath it. By creating a new OPFS instance, all handles pointing to the old instance are guaranteed to never become valid again

Reply:
I suspect it won't quite solve #59 as you can still have that problem within a subdirectory or some such.

Reply:
Correct. I was hoping we'd behave "properly" in at least this situation, but after looking into it more it turns out that the Chromium implementation always re-creates the new OPFS instance at the exact same path. So nevermind on this being an exception.

Basing a handle on a reference would solve this, and we're looking into such a re-architecture, but until that work completes I think we're unfortunately stuck with cross-browser differences on the children of deleted handles

Reply:
So I think specification-wise it's probably easier if we keep the root directory around (so we don't have to meddle with the storage bucket). I think that works because you cannot observe the root directory directly (except for #59 which we have to solve anyway).

Reply:
SGTM. The latest commit adds special casing for the root